state/protocol/protocol_state/epochs/statemachine_test.go

```diff
@@ -0,0 +1,512 @@
+	commitsDB                       *storagemock.EpochCommits
+	globalParams                    *protocolmock.GlobalParams
+	parentState                     *protocol_statemock.KVStoreReader
+	parentEpochState                *flow.RichProtocolStateEntry
+	mutator                         *protocol_statemock.KVStoreMutator
+	happyPathStateMachine           *mock.StateMachine
+	happyPathStateMachineFactory    *mock.StateMachineFactoryMethod
+	fallbackPathStateMachineFactory *mock.StateMachineFactoryMethod
+	candidate                       *flow.Header
+
+	stateMachine *epochs.EpochStateMachine
+}
+
+func (s *EpochStateMachineSuite) SetupTest() {
+	s.epochStateDB = storagemock.NewProtocolState(s.T())
+	s.setupsDB = storagemock.NewEpochSetups(s.T())
+	s.commitsDB = storagemock.NewEpochCommits(s.T())
+	s.globalParams = protocolmock.NewGlobalParams(s.T())
+	s.globalParams.On("EpochCommitSafetyThreshold").Return(uint64(1_000))
+	s.parentState = protocol_statemock.NewKVStoreReader(s.T())
+	s.parentEpochState = unittest.ProtocolStateFixture()
+	s.mutator = protocol_statemock.NewKVStoreMutator(s.T())
+	s.candidate = unittest.BlockHeaderFixture(unittest.HeaderWithView(s.parentEpochState.CurrentEpochSetup.FirstView + 1))
+	s.happyPathStateMachine = mock.NewStateMachine(s.T())
+	s.happyPathStateMachineFactory = mock.NewStateMachineFactoryMethod(s.T())
+	s.fallbackPathStateMachineFactory = mock.NewStateMachineFactoryMethod(s.T())
+
+	s.epochStateDB.On("ByBlockID", mocks.Anything).Return(func(_ flow.Identifier) *flow.RichProtocolStateEntry {
+		return s.parentEpochState
+	}, func(_ flow.Identifier) error {
+		return nil
+	})
+	s.parentState.On("GetEpochStateID").Return(func() flow.Identifier {
+		return s.parentEpochState.ID()
+	})
+
+	s.happyPathStateMachineFactory.On("Execute", s.candidate.View, s.parentEpochState).
+		Return(s.happyPathStateMachine, nil).Once()
+
+	s.happyPathStateMachine.On("ParentState").Return(s.parentEpochState).Maybe()
+
+	var err error
+	s.stateMachine, err = epochs.NewEpochStateMachine(
+		s.candidate.View,
+		s.candidate.ParentID,
+		s.globalParams,
+		s.setupsDB,
+		s.commitsDB,
+		s.epochStateDB,
+		s.parentState,
+		s.mutator,
+		s.happyPathStateMachineFactory.Execute,
+		s.fallbackPathStateMachineFactory.Execute,
+	)
+	require.NoError(s.T(), err)
+}
+
+// TestBuild_NoChanges tests that hierarchical epoch state machine maintains index of epoch states and commits
+// epoch state ID in the KV store even when there were no events to process.
+func (s *EpochStateMachineSuite) TestBuild_NoChanges() {
+	s.happyPathStateMachine.On("ParentState").Return(s.parentEpochState)
+	s.happyPathStateMachine.On("Build").Return(s.parentEpochState.ProtocolStateEntry, s.parentEpochState.ID(), false).Once()
+
+	err := s.stateMachine.ProcessUpdate(nil)
+	require.NoError(s.T(), err)
+
+	indexTxDeferredUpdate := storagemock.NewDeferredDBUpdate(s.T())
+	indexTxDeferredUpdate.On("Execute", mocks.Anything).Return(nil).Once()
```

Comment:
the `Execute` is a general mocked method for all anonymous functions. Since a `DeferredDBUpdate` is a anonymous function we expect call to `Execute` since we pass it in next [line](https://github.com/onflow/flow-go/pull/5616/files/871174eb00fdc1149bbc1c7bafe94f43fa5a6e4c#diff-f20f81e74ce9a22d009f7300d0a42cd62faa7885cfd78ae05b8928ead315f506R104).